### PR TITLE
Use custom exception for warn-as-error instead of user_err + polymorphic maps

### PR DESCRIPTION
**Pull request by SkySkimmer**
This may be a little faster if some tactic code is using warn-as-error
for flow control, but mostly it's because we prefer to avoid user_err.

### CODE REVIEW

**Comment by ejgallego** [comment]
I guess indeed that's better, couple of comments:
- not sure adding yet another `O(n)` printing path is a good thing, don't we have an alternative that doesn't require a linear scan/lookup?
- should maybe the exception be exported (so it can be reified by document managers for example?) Not a big deal.

**Comment by SkySkimmer** [comment]
>not sure adding yet another O(n) printing path is a good thing, don't we have an alternative that doesn't require a linear scan/lookup?

I guess instead of 
~~~ocaml
type v = ..
val printers : (v -> pp option) list ref 
~~~
we could have something like 
~~~ocaml
type _ tag = ..
type v = V : 'a tag * 'a -> v
val printers : polymorphic map from 'a tag to 'a
~~~
polymorphic map kinda like in https://github.com/coq/coq/pull/17145

**Comment by ejgallego** [comment]
IMVHO that would be nice, also for errors.

I'm still hoping to make progress one day on #12602 , which could use such a map, as I'd like for IDEs to have the choice to have a richer structure to handle error messages (and actions).

**Comment by SkySkimmer** [comment]
added some polymorphic map system and used it also for grammar (subsumes https://github.com/coq/coq/pull/17145)
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

**Review by ejgallego — APPROVED** [review]
LGTM , I guess @ppedrot wants to have a look at this too.

Maybe it would be better not to expose `find` but just `find_opt` so we force callers to handle `Not_found` ?

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
@coqbot bench

File: clib/polyMap.ml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  end
+
+  module Map (V:ValueS) = struct
+
+    type v = V : 'a onetag * 'a V.t -> v
+
+    let key t = Obj.Extension_constructor.(id (of_val t))
+    let onekey t = key (tag_of_onetag t)
+
+    module M = Int.Map
```
Comment (SkySkimmer) [inline clib/polyMap.ml:65]: Using Int.Map on the extension constructor feels cleaner than HMap using polymorphic compare and hash

**Comment by ejgallego** [comment]
@ppedrot want to have a look at this?

**Review by ppedrot — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by ejgallego** [comment]
fiat-crypto needs to be cut down.
@coqbot: merge now

**Pull request merged**